codebase/src/compiler/checker-inference.md

```diff
@@ -0,0 +1,245 @@
+case. Now, as before, we look inside `setup`'s type: `(name, age) =>
+...` is the first argument, so its contextual type is from the first
+parameter of `setup`, `register`. `assignmentContextualParameterTypes`
+works for `name` and `age` as in (1).
+
+## Type Parameter Inference
+
+Type parameter inference is quite different from the other two
+techniques. It still infers **types** based on provided **values**,
+but the inferred types don't replace a type annotation. Instead
+they're provided as type arguments to a function, which results in
+instantiating a generic function with some specific type. For example:
+
+``` ts
+declare function setup<T>(config: { initial(): T }): T
+setup({ initial() { return "last" } })
+```
+
+First checks `{ initial() { return "last" } }` to get `{ initial():
+string }`. By matching `T` in `{ initial(): T }` with `string` in `{
+initial(): string }`, it infers that `T` is `string`, making the
+second line the same as if the author had written:
+
+``` ts
+setup<string>({ initial() { return "last" } })
+```
+
+Meaning that the compiler then checks that
+`{ initial() { return "last" } }` is assignable to
+`{ initial(): string }`.
+
+### Walkthrough
+
+Type parameter inference starts off in `inferTypeArguments`, where
+the first step in type parameter inference is to get the type of all
+the arguments to the function whose parameters are being inferred. In
+the above example, the checker says that the type of
+`{ initial() { return "last" } }` is `{ initial(): string }`. This
+type is called the **source** type, since it is the source of
+inferences. It's matched with the parameter type `{ initial(): T }`.
+This is the **target** type -- it contains type parameters which are
+the target of the process.
+
+Type parameter inference is a pairwise walk of the two types, looking
+for type parameters in the target, matching them to corresponding
+types in the source. The type is walked structurally sort of like a tree
+is elsewhere in the compiler.
+
+1. `inferTypes` gets called on each source/target pair with
+   argument=source/parameter=target. There's only one pair here:
+   `{ initial(): string }` and `{ initial(): T }`.
+2. Since both sides are object types, `inferFromProperties` looks
+   through each property of the target and looks for a match in the
+   source. In this case both have the property `initial`.
+3. `initial`'s type is a signature on both sides
+   (`() => T/() => string`), so inference goes to `inferFromSignature`, which
+   recursively infers from the return type.
+4. Now the source/target pair is `T/string`. Since the source is a
+   lone type parameters, we add `string` to the list of candidates for
+   `T`.
+
+Once all the parameters have had `inferTypes` called on them,
+`getInferredTypes` condenses each candidate array to a single type,
+via `getUnionType` in this case. `T`'s candidates array is `[string]`,
+so `getUnionType` immediately returns `string`.
```

Comment:
If you wanna call back to your prior "it's not Hindley-Milner" comment, you can say that this process is kinda like us generating equalities for Hindley-Milner and then solving them at the same time. (Or saying we only accept equalities in the form `T=something else`, where `T` is our goal type parameter). Eh, readers who care will probably realize that on their own anyway, probably no need to mention it.